lib/session.c

```diff
@@ -33,43 +33,58 @@
 #include "session.h"
 #include "traverse.h"
 
+#if HAVE_BTRFS_H
+#include <linux/btrfs.h>
+#endif
+
+#if HAVE_LINUX_FS_H
+#include <linux/fs.h>
+#endif
+
+#ifdef FIDEDUPERANGE
+#define HAVE_FIDEDUPERANGE 1
+#else
+#define HAVE_FIDEDUPERANGE 0
+#endif
+
+#if HAVE_BTRFS_H || HAVE_FIDEDUPERANGE
+#include <sys/ioctl.h>
+#endif
+
 #if HAVE_UNAME
 #include "sys/utsname.h"
+#endif
 
-void rm_session_read_kernel_version(RmSession *session) {
+static gpointer rm_session_read_kernel_version(_UNUSED gpointer arg) {
+    static int version[2] = {-1, -1};
+#if HAVE_UNAME
     struct utsname buf;
-    if(uname(&buf) == -1) {
-        return;
+    if(uname(&buf) != -1 && sscanf(buf.release, "%d.%d.*", &version[0], &version[1]) != EOF) {
+        rm_log_debug_line("Linux kernel version is %d.%d.", version[0], version[1]);
+    } else {
+        rm_log_warning_line("Unable to read Linux kernel version");
     }
-
-    if(sscanf(buf.release, "%d.%d.*", &session->kernel_version[0],
-              &session->kernel_version[1]) == EOF) {
-        session->kernel_version[0] = -1;
-        session->kernel_version[1] = -1;
-        return;
-    }
-
-    rm_log_debug_line("Linux kernel version is %d.%d.",
-                      session->kernel_version[0],
-                      session->kernel_version[1]);
-}
 #else
-void rm_session_read_kernel_version(RmSession *session) {
-    (void)session;
-}
+    rm_log_warning_line(
+        "rmlint was not compiled with ability to read Linux kernel version");
 #endif
+    return version;
+}
 
-bool rm_session_check_kernel_version(RmSession *session, int major, int minor) {
-    int found_major = session->kernel_version[0];
-    int found_minor = session->kernel_version[1];
+bool rm_session_check_kernel_version(int need_major, int need_minor) {
+    static GOnce once = G_ONCE_INIT;
+    g_once (&once, rm_session_read_kernel_version, NULL);
+    int *version = once.retval;
+    int major = version[0];
+    int minor = version[1];
 
-    /* Could not read kernel version: Assume failure on our side. */
-    if(found_major <= 0 && found_minor <= 0) {
```

Comment:
Yes, I saw that you changed that earlier... I should really take courses on version checks. ;)